### ISSUE TITLE
tagged-union narrowing via enum-literal attributes in match/case

### ISSUE THREAD

**Issue by jack-volantautonomy** [astral-sh/ty#2897]
### Summary

We are currently suffering from a lot of false errors being raised with our type aliasing with scenarios like the following:

```py
from typing import Literal, TypeAlias

from pydantic import BaseModel


class _FruitBase(BaseModel):
    id: UUID
    taste: str

class Orange(_FruitBase):
    type: Literal[FruitType.Citrus] = FruitType.Citrus

class Strawberry(_FruitBase):
    type: Literal[FruitType.Berry] = FruitType.Berry
    strawberry_exclusive_params: StrawberryExclusiveParams

CafeFruit: TypeAlias = Orange | Strawberry

def example_method(fruits: list[CafeFruit]):
    for fruit in fruits:
        match fruit.type:
            case FruitType.Berry:  # Should be able to infer fruit: Strawberry
                 foo = fruit.strawberry_exclusive_params  # <--- Issue flags here
```

Which would raise _"error[unresolved-attribute]: Attribute `strawberry_exclusive_params` is not defined on `Orange` in union `Orange | Strawberry`"_

Apologies if this issue has already been raised but we're getting a lot of these being flagged by `ty` but not by any of our other linters that we run in parallel like `pyright` and the `pylance` language server works perfectly with this scenario.

### Version

ty 0.0.18

**Comment by AlexWaygood** [comment]
Thanks for the report! I believe this is a combination of https://github.com/astral-sh/ty/issues/1479 and https://github.com/astral-sh/ty/issues/2192

**Comment by jack-volantautonomy** [comment]
> Thanks for the report! I believe this is a combination of [#1479](https://github.com/astral-sh/ty/issues/1479) and [#2192](https://github.com/astral-sh/ty/issues/2192)

Yes this does look very similar to those 2 issues! I think I'd like to keep this issue open though until this implementation is resolved too as it differs ever so slightly and I just want to make sure this case is checked against still :)

**Comment by carljm** [comment]
> Thanks for the report! I believe this is a combination of [#1479](https://github.com/astral-sh/ty/issues/1479) and [#2192](https://github.com/astral-sh/ty/issues/2192)

Yes -- as well as ensuring that we do this narrowing via match patterns, too.

**Comment by carljm** [comment]
Simplified self-contained repro:

```py
from typing import Literal, TypeAlias

from enum import Enum

class FruitType(Enum):
    Berry = "berry"
    Citrus = "citrus"

class Orange:
    type: Literal[FruitType.Citrus] = FruitType.Citrus

class Strawberry:
    type: Literal[FruitType.Berry] = FruitType.Berry
    strawberry_exclusive_attr: int

CafeFruit: TypeAlias = Orange | Strawberry

def example_method(fruits: list[CafeFruit]):
    for fruit in fruits:
        match fruit.type:
            case FruitType.Berry:  # Should be able to infer fruit: Strawberry
                 foo = fruit.strawberry_exclusive_attr  # <--- Issue flags here
```

https://play.ty.dev/8840906b-3698-44ba-bf5b-1ba5e142c5d7

**Comment by charliermarsh** [comment]
Closed by https://github.com/astral-sh/ruff/pull/25855
